### ISSUE TITLE
Product of `Parameter` and 0 returns `ParameterExpression(0)`

### ISSUE THREAD

**Issue by SimoneGasperini** [Qiskit/qiskit-terra#10243]
### Environment

- **Qiskit Terra version**: 0.22.3
- **Python version**: 3.9.16

### What is happening?

The left/right multiplication of any `Parameter` with 0 returns the object `ParameterExpression(0)`.

### How can we reproduce the issue?

```python
from qiskit.circuit import Parameter

0 * Parameter('x') == 0  # False
Parameter('x') * 0 == 0  # False
```

### What should happen?

The left/right multiplication of any `Parameter` with 0 should return 0 in order to make `True` both the comparisons above. This should work also when we use the complex number `0 + 0j`.

**Comment by jakelishman** [comment]
This is expected behaviour, because it maintains the invariant that a `ParamaterExpression` with unbound paramaters can still accept `.bind` calls for the inner symbols, even if they've been cancelled out. This makes it programmatically easier to deal with, as a series of expressions involving the same paramaters can all be bound with a dictionary defining all of them without complaint.

**Comment by jakelishman** [comment]
It's more up in the air whether the equality should hold, but equality in symbolic expressions is a tricky question to pin down exactly (and `ParamaterExpression` is currently very badly behaving in the related `__hash__` function, which isn't great).

**Comment by SimoneGasperini** [comment]
Thank you @jakelishman for the clarification. I close this as resolved.

**Comment by jakelishman** [comment]
Just for posterity, I don't think I gave the best answer the first time round, so let me try again.

There's no inherently correct answer as to whether a binary operation `A * B` should yield a value of type `A` or of type `B`.
- Some languages (e.g. Rust) say the `*` operation can only be valid between objects of the _same_ type, like `A * A -> A`.
- Some (e.g. C) define a set of "promotion rules" that choose a third type `C` that the `A` value and the `B` value are both cast to, and the result is of type `C` (often type `C` is one of `A` or `B`).
- Some (e.g. Python and C++) the `*` operation to be implemented with an arbitrary signature `A * B -> C` for any A, B and C.

Python allows us to be in the latter camp, but in practice, it's usually quite surprising if the logical return type isn't either of the input types.  Now we're left with the question of "is it better to return a `float` or a `ParameterExpession`?".  In most duck-typed uses, `ParameterExpression` is structural supertype of `float`, so we return `ParameterExpression` because then you can rely on `(x * expr).assign(...)` working as expected - if we returned `float` in some cases, you'd have to check the output.

We can universally define `float * ParameterExpression -> ParameterExpression` for all `float` values, so there's no worry that _some_ floats would cause the type of the return to change.  That's a sort of "dependent typing", and typically it's an easy way to get _super_ weird errors that are delocalised from the actual place the typing became different from what you expect.

**Comment by SimoneGasperini** [comment]
Thank you again! Very clear answer!!
